File: android/DFM/src/main/java/com/dontflymoney/entities/Move.java
package com.dontflymoney.entities;

import com.dontflymoney.api.Request;
import com.dontflymoney.viewhelper.DateTime;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class Move
{
	public Move()
	{
		Details = new ArrayList<>();
		Date = Calendar.getInstance();
	}

    public int Id;
	public String Description;
    public Calendar Date;
    public String Category;
    public String PrimaryAccount;
    public String OtherAccount;
    public int Nature;

    public boolean isDetailed;
    public double Value;
    public List<Detail> Details;



    public void Add(String description, int amount, double value)
    {
    	Detail detail = new Detail();
    	
    	detail.Description = description;
    	detail.Amount = amount;
    	detail.Value = value;
    	
    	Details.add(detail);
    }

	public void Remove(String description, int amount, double value)
	{
		for (Detail detail : Details)
		{
			if (detail.Equals(description, amount, value))
			{
				Details.remove(detail);
				return;
			}
		}
	}
	
	
	
	public int getDay()
	{
		return Date.get(Calendar.DAY_OF_MONTH);
	}
	
	public int getMonth()
	{
		return Date.get(Calendar.MONTH);
	}
	
	public int getYear()
	{
		return Date.get(Calendar.YEAR);
	}

	public void setParameters(Request request)
	{
        request.AddParameter("ID", Id);
        request.AddParameter("Description", Description);
        request.AddParameter("Date.Year", Date.get(Calendar.YEAR));
        request.AddParameter("Date.Month", Date.get(Calendar.MONTH) + 1);
        request.AddParameter("Date.Day", Date.get(Calendar.DAY_OF_MONTH));
		request.AddParameter("Category", Category);
		request.AddParameter("PrimaryAccount", PrimaryAccount);
		request.AddParameter("OtherAccount", OtherAccount);
		request.AddParameter("Nature", Nature);

        if (isDetailed)
        {
            for (Detail detail : Details)
            {
                int position = Details.lastIndexOf(detail);

                request.AddParameter("DetailList[" + position + "].Description", detail.Description);
                request.AddParameter("DetailList[" + position + "].Amount", detail.Amount);
                request.AddParameter("DetailList[" + position + "].Value", detail.Value);
            }
        }
        else
        {
            request.AddParameter("Value", Value);
        }
	}

	public String DateString() {
		SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault());

		return formatter.format(Date.getTime());
	}




    public void SetData(JSONObject moveToEdit) throws JSONException
    {
        Id = moveToEdit.getInt("ID");
        Description = moveToEdit.getString("Description");
        Date = DateTime.getCalendar(moveToEdit.getJSONObject("Date"));
        Category = moveToEdit.getString("Category");
        OtherAccount = moveToEdit.getString("OtherAccount");
        Nature = moveToEdit.getInt("Nature");

        if (moveToEdit.has("Value") && !moveToEdit.get("Value").equals(null))
        {
            Value = moveToEdit.getDouble("Value");
        }

        if (moveToEdit.has("DetailList"))
        {
            JSONArray detailList = moveToEdit.getJSONArray("DetailList");

            isDetailed = detailList.length() > 0;

            for (int d = 0; d < detailList.length(); d++)
            {
                JSONObject detail = detailList.getJSONObject(d);
                Add(detail.getString("Description"), detail.getInt("Amount"), detail.getDouble("Value"));
            }
        }
    }
	
	
}

File: android/DFM/src/main/java/com/dontflymoney/entities/Move.kt
package com.dontflymoney.entities

import com.dontflymoney.activityObjects.MovesCreateStatic
import com.dontflymoney.api.InternalRequest
import com.dontflymoney.viewhelper.DateTime
import org.json.JSONException
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*

class Move {

    private var Id: Int = 0
    var Description: String? = null
    var Date: Calendar = Calendar.getInstance()

    var Nature: Nature? = null

    var Category: String? = null
    var AccountOut: String? = null
    var AccountIn: String? = null

    var isDetailed: Boolean = false
    var Value: Double = 0.toDouble()
    var Details: MutableList<Detail> = ArrayList<Detail>()


    private fun getNature(number: Int?): Nature {
        return com.dontflymoney.entities.Nature.GetNature(number)
    }


    fun SetNature(number: String) {
        SetNature(Integer.parseInt(number))
    }

    fun SetNature(number: Int?) {
        Nature = getNature(number)
    }

    fun Add(description: String, amount: Int, value: Double) {
        val detail = Detail()

        detail.Description = description
        detail.Amount = amount
        detail.Value = value

        Details.add(detail)
    }

    fun Remove(description: String?, amount: Int, value: Double) {
        for (detail in Details) {
            if (detail.Equals(description, amount, value)) {
                Details.remove(detail)
                return
            }
        }
    }


    val day: Int
        get() = Date.get(Calendar.DAY_OF_MONTH)

    val month: Int
        get() = Date.get(Calendar.MONTH)

    val year: Int
        get() = Date.get(Calendar.YEAR)

    fun setParameters(request: InternalRequest<MovesCreateStatic>) {
        request.AddParameter("ID", Id)
        request.AddParameter("Description", Description)

        request.AddParameter("Date.Year", Date.get(Calendar.YEAR))
        request.AddParameter("Date.Month", Date.get(Calendar.MONTH) + 1)
        request.AddParameter("Date.Day", Date.get(Calendar.DAY_OF_MONTH))

        request.AddParameter("Nature", Nature)

        request.AddParameter("Category", Category)
        request.AddParameter("AccountOutUrl", AccountOut)
        request.AddParameter("AccountInUrl", AccountIn)

        if (isDetailed) {
            for (detail in Details) {
                val position = Details.lastIndexOf(detail)

                request.AddParameter("DetailList[$position].Description", detail.Description!!)
                request.AddParameter("DetailList[$position].Amount", detail.Amount)
                request.AddParameter("DetailList[$position].Value", detail.Value)
            }
        } else {
            request.AddParameter("Value", Value)
        }
    }

    fun DateString(): String {
        val formatter = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())

        return formatter.format(Date.time)
    }


    @Throws(JSONException::class)
    fun SetData(move: JSONObject, activityAccountUrl: String) {
        Id = move.getInt("ID")

        if (Id == 0) {
            AccountOut = activityAccountUrl
        } else {
            setEditMoveData(move)
        }
    }

    @Throws(JSONException::class)
    private fun setEditMoveData(move: JSONObject) {
        Description = move.getString("Description")
        Date = DateTime.getCalendar(move.getJSONObject("Date"))
        Category = move.getString("Category")
        AccountOut = move.getString("AccountOutUrl")
        AccountIn = move.getString("AccountInUrl")
        Nature = getNature(move.getInt("Nature"))

        if (move.has("Value") && !move.isNull("Value")) {
            Value = move.getDouble("Value")
        }

        if (move.has("DetailList")) {
            val detailList = move.getJSONArray("DetailList")

            isDetailed = detailList.length() > 0

            for (d in 0..detailList.length() - 1) {
                val detail = detailList.getJSONObject(d)
                Add(detail.getString("Description"), detail.getInt("Amount"), detail.getDouble("Value"))
            }
        }
    }


}
